dlib/dnn/layers.h

```diff
@@ -594,9 +594,334 @@ namespace dlib
+                << " scale_y='"<<_scale_y<<"'"
+                << " scale_x='"<<_scale_x<<"'"
+                << " method='>\n"<<_method<<"'";
+            out << mat(item.params);
+            out << "</upsample>";
+        }
+
+    private:
+
+        resizable_tensor params;
+
+        tt::tensor_upsample ups;
+    };
+
+    template <
+        int scale_y,
+        int scale_x,
+        unsigned char method,
+        typename SUBNET
+        >
+    using upsample = add_layer<upsample_<scale_y,scale_x,method>, SUBNET>;
+
 // ----------------------------------------------------------------------------------------
 
     template <
+        int _padding_y,
+        int _padding_x,
+        unsigned char _method
+        >
+    class spatialpadding_
+    {
+    public:
+
+        static_assert(_padding_y > 0, "The padding must be > 0");
+        static_assert(_padding_x > 0, "The padding must be > 0");
+        static_assert(0 <= _method && _method < 3, "The method must be between 0 and 2");
+
+        spatialpadding_(
+        )  
+        {
+        }
+
+        long padding_y() const { return _padding_y; }
+        long padding_x() const { return _padding_x; }
+        unsigned char method() const { return _method; }
+
+
+        inline point map_output_to_input (
+            point p
+        ) const
+        {
+            p.x() = p.x() + 2 * padding_x();
+            p.y() = p.y() + 2 * padding_y();
+            return p;
+        }
+
+        inline point map_input_to_output (
+            point p
+        ) const
+        {
+            p.x() = p.x() - 2 * padding_x();
+            p.y() = p.y() - 2 * padding_y();
+            return p;
+        }
+
+        spatialpadding_ (
+            const spatialpadding_& item
+        )  
+        {
+            (void)item;
+            // this->conv is non-copyable and basically stateless, so we have to write our
+            // own copy to avoid trying to copy it and getting an error.
+        }
+
+        spatialpadding_& operator= (
+            const spatialpadding_& item
+        )
+        {
+            if (this == &item)
+                return *this;
+
+            // this->conv is non-copyable and basically stateless, so we have to write our
+            // own copy to avoid trying to copy it and getting an error.
+            return *this;
+        }
+
+        template <typename SUBNET>
+        void setup (const SUBNET& sub)
+        {
+            (void)sub;
+            params.set_size(0);
+        }
+
+        template <typename SUBNET>
+        void forward(const SUBNET& sub, resizable_tensor& output)
+        {
+            unsigned int gnr = sub.get_output().nr() + 2 * padding_y();
+            unsigned int gnc = sub.get_output().nc() + 2 * padding_x();
+            unsigned int gnsamps = sub.get_output().num_samples();
+            unsigned int gk = sub.get_output().k();
+            output.set_size(gnsamps,gk,gnr,gnc);
+            output = 0;
```

Comment:
Same question here about why pad.forward() needs to have zeroed inputs.  Can't pad.forward() do an assignment rather than an accumulate to the output?